### PR TITLE
Redesign isOpSupported()

### PR DESCRIPTION
**Pull request by jfix71**
*Description*: This PR redesigns `isOpSupported()` to allow a backend to specify what ops it supports based on the input and output types of Nodes. It updates all of the open source backends we have.

*Testing*: All tests are passing. I've also done my best to make sure bigger tests (e.g. those in `tests/images/run.sh`) also still work.

*Documentation*: Updated.

Fixes https://github.com/pytorch/glow/issues/2194

### CODE REVIEW

**Comment by jackm321** [comment]
@jfix71 I'm glad to see this move in the direction of considering all inputs and outputs. I think the cost function is cool but I'm also a little unclear about how it would be used in practice. Is it imagined that the costs of various operators would just be compared within the same backend for different operators? This doesn't sound likely so then how will we ensure that these costs are defined so that they are comparable across backends when each backend defines it's own cost for each operator?

**Comment by jfix71** [comment]
> Is it imagined that the costs of various operators would just be compared within the same backend for different operators? This doesn't sound likely so then how will we ensure that these costs are defined so that they are comparable across backends when each backend defines it's own cost for each operator?

Good questions! I was moving in this direction as suggested in https://github.com/pytorch/glow/issues/2194#issuecomment-449191002.

Overall yeah, I think it's more useful across different backends, both from the perspective of partitioning and from ONNXIFI determining where to execute a given function. If we wanted the cost to represent execution speed, I think we could just define these based on wall time, as we have an expectation of how long each operator would take to execute.

For the same backend, perhaps it could be useful to know for deciding where/when to partition a function across multiple of the same backend. As a really trivial (and perhaps unhelpful 😛) example, assume individually `Conv` and `Relu` cost "1", and a fused `ConvRelu` also costs "1" since `Relu` is essentially free on some accelerators when fused to `Conv`. So this logic lets you know it's a bad idea to partition in between them.

**Comment by jackm321** [comment]
@jfix71 I don't know a lot about the hardware but estimating wallclock times for various inputs and outputs sounds like it would be difficult to get to be correct and approximations may miss things like the ConvRelu example if they're not accurate enough to discern that Conv and ConvRelu are the same cost.

Also for the ConvRelu fusion example, does the partitioner get an optimized graph or not? Because if it does it seems like it would be better just to have an optimization pass (or prelowering maybe?) that does fusions like this for sure on backends where it's good to do that and then the partitioner doesn't have to worry about whether it should or shouldn't make a cut there in the graph based on cost because it will just be a single fused node at that point.

Overall I guess my concern is that this cost function is nice because the idea behind `UnsupportedCostVal = std::numeric_limits<CostVal>::max()` but in practice I'm not sure if the values between `1` and `UnsupportedCostVal` can actually be reliably generated and used by the partitioner and/or onnxifi so this might be overly general. Not at all sure if this is true or not though 😛

**Comment by jfix71** [comment]
> I don't know a lot about the hardware but estimating wallclock times for various inputs and outputs sounds like it would be difficult to get to be correct

@jackm321 I agree it would be difficult for an entire graph, but I don't think estimating a single operator's costs would be too difficult. It would be the responsibility of the higher level partitioner to understand the backend's architecture and the graph as a whole to make decisions on when to partition. Perhaps backends would need a `getSubgraphCost()` which would call into `getOpCost()`. Overall though I'll defer here to @gcatron and others working on the partitioner/runtime that have a better idea of where the partitioner is going/what info it needs/how it should get that info.

Also, another place this could be useful for a single backend is in the GraphOptimizer. Right now backends have no control over whether an optimization should take place (https://github.com/pytorch/glow/issues/1641). One solution here could be for the GraphOptimizer to query the backend to get the costs of the original subgraph and the new subgraph to decide whether to perform the opti. Similar to the partitioner, we could get each cost using something like `getSubgraphCost()`, calling into `getOpCost()`.

File: include/glow/ExecutionEngine/ExecutionEngine.h [hunk 1]
```diff
@@ -67,11 +67,17 @@ class ExecutionEngine final {
...
+  /// \returns a cost associated with executing the provided \p opKind with \p
+  /// inputTypes and \p outputTypes on the underlying backend.
+  CostVal getOpCost(Kinded::Kind opKind, llvm::ArrayRef<TypeRef> inputTypes,
+                    llvm::ArrayRef<TypeRef> outputTypes) const {
+    return backend_->getOpCost(opKind, inputTypes, outputTypes);
   }
 
+  /// \returns whether the cost associated with executing Node \p N on the
+  /// underlying backend.
+  CostVal getOpCost(const Node &N) const { return backend_->getOpCost(N); }
```
Comment (zakk0610) [inline include/glow/ExecutionEngine/ExecutionEngine.h]: I think maybe this function need to be `virtual` for backend to provide more accurate estimation.
because the cost is depend on operator's parameter., 
For example, if the input size is too small, the CPU maybe faster than the ASIC.

Comment (jfix71) [reply include/glow/ExecutionEngine/ExecutionEngine.h]: The actual logic for computing an op's cost is contained in a virtual `getOpCost()`:

https://github.com/pytorch/glow/blob/0c01a667b8fb26f7752ae7ed63c4b6d134ed729c/include/glow/Backends/Backend.h#L78-L82

This function here in ExecutionEngine just a helper method. It forwards to the backend's helper `getOpCost(Node &N)`, which then calls into each backend's implementation of `getOpCost()`.

Comment (zakk0610) [reply include/glow/ExecutionEngine/ExecutionEngine.h]: Thanks, I was missed the `TypeRef` can represents a type of tensor. 
Sorry to bother you.

Comment (jfix71) [reply include/glow/ExecutionEngine/ExecutionEngine.h]: No bother at all, thanks for checking out the PR! We are happy to hear any questions or insights from everyone contributing to the project.

File: include/glow/Graph/Node.h [hunk 2]
```diff
@@ -145,6 +145,9 @@ class Node : public Named,
 
   ~Node() { releaseUsers(); }
 
+  /// \returns an array ref of the types of the node.
```
Comment (rdzhabarov) [inline include/glow/Graph/Node.h]: based on the comment it's not clear whether it's input/output or both.

File: lib/Backends/Backends.cpp [hunk 3]
```diff
@@ -62,3 +62,14 @@ Backend *glow::createBackend(BackendKind backendKind) {
   // always covers all possible values.
   llvm_unreachable("unreachable");
 }
+
+CostVal Backend::getOpCost(const Node &N) const {
+  // Gather the input types into a vector. Note that we do not need to do so for
+  // the result types because they are already contained in the node.
+  std::vector<TypeRef> inputTypeRefs;
```
Comment (rdzhabarov) [inline lib/Backends/Backends.cpp]: nit: but when we know size in advance, just allocate it right away in vector.

File: lib/Backends/CPU/CPUBackend.cpp [hunk 4]
```diff
@@ -168,13 +172,13 @@ bool CPUBackend::isOpSupported(Kinded::Kind opKind, ElemKind elementTy) const {
     case Kinded::Kind::TileNodeKind:
     case Kinded::Kind::TopKNodeKind:
     case Kinded::Kind::TransposeNodeKind:
-      return true;
+      return 1;
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CPUBackend.cpp]: can you make this a named const (MinimalCostVal, there might be a better name)?

File: lib/Backends/CPU/CPUBackend.cpp [hunk 5]
```diff
@@ -135,7 +135,11 @@ void CPUBackend::save(Function *F, llvm::StringRef outputDir,
   BundleSaver(IR.get()).save(tgt, outputDir, networkName);
 }
 
-bool CPUBackend::isOpSupported(Kinded::Kind opKind, ElemKind elementTy) const {
+CostVal CPUBackend::getOpCost(Kinded::Kind opKind,
+                              llvm::ArrayRef<TypeRef> inputTypes,
+                              llvm::ArrayRef<TypeRef> outputTypes) const {
+  // TEMPORARY: Will remove once not a WIP. This is just so things compile.
+  auto elementTy = inputTypes[0]->getElementType();
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CPUBackend.cpp]: Do you plan on handling int16 and fp32 in the follow up?

File: lib/Backends/CPU/CPUBackend.cpp (hunk 4 above)
Comment (jackm321) [reply lib/Backends/CPU/CPUBackend.cpp]: Yes I think this would help with clarity to have a name for this value. Not sure if minimal is right though because I guess somethings may be 0 if they are "free" like the Relu in ConvRelu example?

Comment (rdzhabarov) [reply lib/Backends/CPU/CPUBackend.cpp]: well it depends, on CPU if not fused `relu` is not "free".

File: lib/Backends/CPU/CPUBackend.cpp (hunk 5 above)
Comment (jfix71) [reply lib/Backends/CPU/CPUBackend.cpp]: Yep! I will update the CPU and OpenCL backend's versions too in this PR. I first just wanted to make sure everyone agreed the Interpreter's `getOpCost()`/the general approach seemed reasonable.

File: include/glow/Graph/Node.h (hunk 2 above)
Comment (jfix71) [reply include/glow/Graph/Node.h]: Done

File: lib/Backends/Backends.cpp (hunk 3 above)
Comment (jfix71) [reply lib/Backends/Backends.cpp]: Done

File: lib/Backends/CPU/CPUBackend.cpp (hunk 4 above)
Comment (jfix71) [reply lib/Backends/CPU/CPUBackend.cpp]: I added `MinimumUnitCostVal` and replaced all `1`s with it. If we keep it I'm thinking backends would use this minimum unit cost to define other costs, i.e. as some multiple of that. It seems a little silly since it is just `1` and I'm not sure we would ever change it from `1`, but perhaps this preferable to just specifying `1`? Let me know what you think.

I also added `FreeCostVal = 0` to differentiate from the minimum, however it's not used anywhere, so not sure it's worth including for now.

Comment (rdzhabarov) [reply lib/Backends/CPU/CPUBackend.cpp]: > It seems a little silly since it is just 1 and I'm not sure we would ever change it from 1, but perhaps this preferable to just specifying 1?

hopefully name should give slightly more context than plain '1'.

> I also added FreeCostVal = 0 to differentiate from the minimum, however it's not used anywhere, so not sure it's worth including for now.

up to you, we can introduce that once there is a need.

Comment (jackm321) [reply lib/Backends/CPU/CPUBackend.cpp]: >  It seems a little silly since it is just 1 and I'm not sure we would ever change it from 1, but perhaps this preferable to just specifying 1?

@jfix71 I don't think it's too silly, I imagine it would give enough context to make it pretty easy to understand what's being expressed

File: include/glow/Backends/Backend.h [hunk 6]
```diff
@@ -71,9 +71,14 @@ class Backend {
...
-  /// the given \p elementTy element type.
-  virtual bool isOpSupported(Kinded::Kind opKind, ElemKind elementTy) const = 0;
+  /// \returns whether the provided \p NI is supported by the backend.
+  virtual bool isOpSupported(Function *F, const NodeInfo &NI) const = 0;
+
+  /// \returns whether the provided \p N is supported by the backend. Gathers
+  /// info from \p N calls into the virtual \p ref isOpSupported().
+  bool isOpSupported(Function *F, const Node &N) const;
+
+  bool isOpSupportedViaLowering(Function *F, const NodeInfo &NI) const;
```
Comment (jackm321) [inline include/glow/Backends/Backend.h]: doxygen comment.

File: include/glow/Optimizer/Optimizer.h [hunk 7]
```diff
@@ -43,6 +43,12 @@ void optimize(Function *F, CompilationMode mode);
 /// operators.
 void lower(Function *F, const Backend &B);
 
+/// \returns a pair of a bool and a vector of the component NodeInfo of nodes
+/// that would be lowered from NI. If no nodes would be lowered from this node,
+/// then the bool is set to false.
```
Comment (jackm321) [inline include/glow/Optimizer/Optimizer.h]: Why is the bool needed? Why not just return the empty vector?

File: include/glow/Graph/Node.h [hunk 8]
```diff
@@ -185,6 +191,35 @@ class NodeWalker {
...
+/// Helper struct to hold info of a Node, containing its \p opKind, \p inTypes,
+/// and \p outTypes
+struct NodeInfo {
+private:
+  Kinded::Kind opKind_;
+  llvm::ArrayRef<TypeRef> inTypes_;
+  llvm::ArrayRef<TypeRef> outTypes_;
+
+public:
+  NodeInfo(Kinded::Kind opKind, llvm::ArrayRef<TypeRef> inTypes,
```
Comment (jackm321) [inline include/glow/Graph/Node.h]: Maybe also add a constructor that takes const Node&?

File: lib/Backends/Backends.cpp [hunk 9]
```diff
@@ -62,3 +62,27 @@ Backend *glow::createBackend(BackendKind backendKind) {
...
+bool Backend::isOpSupportedViaLowering(Function *F, const NodeInfo &NI) const {
+  std::pair<bool, std::vector<NodeInfo>> loweredNodeInfos =
+      getLoweredComponentInfo(F, NI);
+
+  // If the node cannot be lowered then return false.
+  if (!loweredNodeInfos.first) {
+    return false;
+  }
+
+  for (const NodeInfo &NI : loweredNodeInfos.second) {
```
Comment (jackm321) [inline lib/Backends/Backends.cpp]: Could you use a fresh name here instead of aliasing the parameter?

File: lib/Optimizer/Lower.cpp [hunk 10]
```diff
@@ -694,6 +694,141 @@ static void lowerTileNode(Function *F, const TileNode &TN) {
   TN.getResult().replaceAllUsesOfWith(IN);
 }
 
+std::pair<bool, std::vector<NodeInfo>>
+glow::getLoweredComponentInfo(Function *F, const NodeInfo &NI) {
```
Comment (jackm321) [inline lib/Optimizer/Lower.cpp]: Add comments about what this does and to keep this in sync with actual lowering code.
Can we add tests to check that this is consistent with lowering code?

File: include/glow/Optimizer/Optimizer.h (hunk 7 above)
Comment (jfix71) [reply include/glow/Optimizer/Optimizer.h]: Some nodes are lowered by replacing them with their input, e.g. the `RegressionNode` and `AddGradNode`. So we want to say that the Node can be lowered and that it is lowered into nothing.

File: lib/Optimizer/Lower.cpp (hunk 10 above)
Comment (jfix71) [reply lib/Optimizer/Lower.cpp]: There are some comments in `Optimizer.h` -- I added a little more to it.

It shouldn't be too bad to create a `LowerTest.cpp`, which creates nodes and then calls into both `lower()` and `getLoweredComponentInfo()` and makes sure they match.

**Comment by jfix71** [comment]
I've added a few tests in an initial `LowerTest`. It turned out pretty clean, and so should be pretty easy to add new tests for.

I'm worried about the size of this PR at this point. It's really big and this is before I've finished implementing `getLoweredComponentInfo()` or `OpenCL::/CPU::isOpSupported()` or `LowerTest`. Unfortunately I think they make sense/need to land together, though... 😞Any suggestions on how to split it would be welcome!

File: include/glow/Backends/Backend.h [hunk 11]
```diff
@@ -71,9 +71,12 @@ class Backend {
...
   /// @}
 
-  /// \returns true if backend supports given kind of operation with
-  /// the given \p elementTy element type.
-  virtual bool isOpSupported(Kinded::Kind opKind, ElemKind elementTy) const = 0;
+  /// \returns whether the provided \p NI is supported by the backend.
+  virtual bool isOpSupported(Function *F, const NodeInfo &NI) const = 0;
+
+  /// \returns whether the provided \p NI is supported by the backend via
+  /// lowering the \p NI into other NodeInfos, matching the lowering logic.
```
Comment (rdzhabarov) [inline include/glow/Backends/Backend.h]: It would worthwhile to mention here that if node cannot be lowered than it's not `isOpSupportedViaLowering` even though the node itself could be supported otherwise.

File: lib/Graph/Nodes.cpp [hunk 12]
```diff
@@ -296,8 +296,13 @@ static bool verifyRelu(NodeValue result, NodeValue input) {
 
 static bool verifyRegression(NodeValue src, NodeValue dest,
                              NodeValue expected) {
-  return checkSameType(src, dest, dest.getNode()) &&
-         checkSameType(dest, expected, dest.getNode());
+  const Node *parent = dest.getNode();
```
Comment (rdzhabarov) [inline lib/Graph/Nodes.cpp]: this seems unrelated to the PR.

Comment (jfix71) [reply lib/Graph/Nodes.cpp]: It was actually required for MLTest's `testFindPixelRegression` test to pass. It quantizes a regression node, which is eventually lowered into nothing during inference but is there for training. I could move it to another PR if desired.

**Comment by jfix71** [comment]
Now that https://github.com/pytorch/glow/pull/2311 has landed I will remove all the `isOpSupportedViaLowering()`, `getLoweredComponentInfoI()`, etc. stuff. I think the only thing left to do here then will be to update the `isOpSupported()` for the OpenCL and CPU backends.

**Comment by jackm321** [comment]
@jfix71 are you planning on using the details of NodeInfo for `CPUBackend::isOpSupported` and `OCLBackend::isOpSupported`? What's here looks good to me, just not sure if you're planning on adding stuff or not.

**Comment by jfix71** [comment]
@jackm321 Yeah, still need to update `isOpSupported()` for the CPU and OpenCL backends, I just haven't had the cycles to make progress here. Additionally the Interpreter backend still needs some work too -- locally I've fixed a few bugs in what's already up here, but still need to add support for all the gradient Nodes that are missing. I've made some updates locally but haven't pushed them up yet.

**Comment by jfix71** [comment]
Update: Interpreter `isOpSupported()` is now complete and passes all tests. Moving on to updating CPU and OpenCL backends next.

**Comment by jfix71** [comment]
OK! This is finally ready.

FYI: I needed to fix a few tests. This is mostly no big deal. There is one test however which I disabled for the OpenCL backend -- the `end2endInt16` from QuantizationTest. OpenCL has no `Int16QTy` support -- the test previously passed because the quantizer basically just skipped over quantizing the whole thing, and so the results compared to the Interpreter passed. This isn't great -- I've filed an issue about it: https://github.com/pytorch/glow/issues/2464

File: tests/unittests/QuantizationTest.cpp [hunk 13]
```diff
@@ -1708,33 +1711,37 @@ INSTANTIATE_TEST_CASE_P(Interpreter, Quantization,
...
 #ifdef GLOW_WITH_OPENCL
 INSTANTIATE_TEST_CASE_P(
-    OpenCL, Operator,
-    ::testing::Combine(::testing::Values(BackendKind::Interpreter,
-                                         BackendKind::Interpreter),
+    InterpProfOpenCLQuant, Operator,
+    ::testing::Combine(::testing::Values(BackendKind::Interpreter),
                        ::testing::Values(BackendKind::OpenCL)));
-#endif // GLOW_WITH_CPU
+#endif // GLOW_WITH_OPENCL
```
Comment (jackm321) [inline tests/unittests/QuantizationTest.cpp]: nice

File: lib/Backends/CPU/CPUBackend.cpp [hunk 14]
```diff
@@ -160,54 +160,348 @@ void CPUBackend::save(Function *F, llvm::StringRef outputDir,
...
+  // Note: For brevity below, "X ==> Y, Z" signifes that Node X is IRGen'd into
+  // Instructions Y and Z.
+
+  switch (NI.getKind()) {
+  case Kinded::Kind::AddNodeKind:
+  case Kinded::Kind::SubNodeKind:
+  case Kinded::Kind::MulNodeKind:
+  case Kinded::Kind::MaxNodeKind:
+  case Kinded::Kind::MinNodeKind:
+    // Note: Int8QTy Log and Tanh are IRGen'd into a lookup table.
```
[the remark this answers is not in the document]
Comment (jfix71) [reply lib/Backends/CPU/CPUBackend.cpp]: This is how `clang-format` indents comments inside switch statements 🤷‍♂️

File: include/glow/Graph/Node.h [hunk 15]
```diff
@@ -185,6 +187,98 @@ class NodeWalker {
...
+    for (const ElemKind elemKind : allowedElemKinds) {
+      if (allSameElemKind(elemKind, inTypes_, ignoreIn) &&
+          allSameElemKind(elemKind, outTypes_, ignoreOut)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  std::string getDebugDesc() const {
```
[the remark this answers is not in the document]
Comment (jfix71) [reply include/glow/Graph/Node.h]: Not used but useful when debugging so I'd like to keep it.

**Comment by jfix71** [comment]
@jackm321 Thanks for the review! Will merge once tests pass to avoid conflicts.
